src/emqx_delayed_publish.erl

```diff
@@ -39,37 +39,20 @@
 %% Plugin callbacks
 %%------------------------------------------------------------------------------
 
--spec(load() -> ok).
-load() ->
-    emqx:hook('message.publish', {?MODULE, on_message_publish, []}).
-
-on_message_publish(Msg = #message{id = Id, topic = <<"$delayed/", Topic/binary>>, timestamp = Ts}) ->
-    [Delay, Topic1] = binary:split(Topic, <<"/">>),
-    PubAt = case binary_to_integer(Delay) of
-                Interval when Interval < ?MAX_INTERVAL ->
-                    Interval + emqx_time:now_secs(Ts);
-                Timestamp ->
-                    %% Check malicious timestamp?
-                    case (Timestamp - emqx_time:now_secs(Ts)) > ?MAX_INTERVAL of
-                        true  -> error(invalid_delayed_timestamp);
-                        false -> Timestamp
-                    end
-            end,
-    PubMsg = Msg#message{topic = Topic1},
-    ok = store(#delayed_message{key = {PubAt, delayed_mid(Id)}, msg = PubMsg}),
-    {stop, PubMsg};
-
-on_message_publish(Msg) ->
-    {ok, Msg}.
+delay_publish(Msg) ->
+    delay_publish(Msg, undefined).
+
+delay_publish(Msg = #message{id = Id, timestamp = Ts, headers = #{'Will-Delay-Interval' := Interval}}, ClientId) ->
+    PubAt = pub_at(Ts, Interval),
+    store(#delayed_message{key = {PubAt, delayed_mid(Id)}, client_id = ClientId, msg = Msg}).
+
+cancel_publish(ClientId) ->
```

Comment:
Not `ClientID`, should be `ID`. User specifies a message id for its delayed message when calling this API.